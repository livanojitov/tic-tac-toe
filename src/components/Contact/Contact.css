.contact{
  width: 95%;
  margin: 0 auto
}

.contact .me, .contact .availability{
  color: green;
  font-weight: normal;
}

.contact span, .about span{
  font-weight: bold;
}