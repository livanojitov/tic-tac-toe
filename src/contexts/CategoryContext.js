import React, { createContext } from "react";
import useFetchData from "../hooks/useFetchData";

export const CategoryContext = createContext();

const CategoryContextProvider = ({ children }) => {
  let { data: categories, error } = useFetchData("categories");

  if (error) {
    return (
      <div className="error" data-testid="error">
        {error}
      </div>
    );
  }

  return (
    <CategoryContext.Provider value={{ categories }}>
      {children}
    </CategoryContext.Provider>
  );
};

export default CategoryContextProvider;
