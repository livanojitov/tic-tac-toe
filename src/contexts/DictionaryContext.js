import React, { createContext } from "react";
import useFetchData from "../hooks/useFetchData";
import useLanguage from "../hooks/useLanguage";
import * as constants from "../Constants";

export const DictionaryContext = createContext();

const DictionaryContextProvider = ({ children }) => {
  let { data: dictionary, error } = useFetchData("dictionary");

  const [language, changeLanguageHandler] = useLanguage(constants.ENGLISH);

  if (error) {
    return (
      <div className="error" data-testid="error">
        {error}
      </div>
    );
  }

  dictionary = dictionary[language];

  return (
    <DictionaryContext.Provider value={{ dictionary, changeLanguageHandler }}>
      {children}
    </DictionaryContext.Provider>
  );
};

export default DictionaryContextProvider;
