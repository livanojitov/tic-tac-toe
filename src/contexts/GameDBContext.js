import React, { createContext } from "react";
import useLocalStorage from "../hooks/useLocalStorage";
export const GameDBContext = createContext();

const GameDBContextProvider = ({ children }) => {
  const { games, dispatch } = useLocalStorage("tictactoe", []);

  const getGame = (id) => {
    if (games.length > 0) {
      return games.filter((game) => game.id === id)[0];
    } else {
      return false;
    }
  };

  return (
    <GameDBContext.Provider value={{ games, dispatch, getGame }}>
      {children}
    </GameDBContext.Provider>
  );
};

export default GameDBContextProvider;
