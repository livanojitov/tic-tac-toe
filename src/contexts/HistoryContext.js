import React, { createContext, useState } from 'react';

export const HistoryContext = createContext();

const HistoryContextProvider = (props) => {

  const [history, setHistory] = useState(0);

  const changeHistory = (history) => {
    setHistory(history);
  }

  return (
    <HistoryContext.Provider value={{ history, changeHistory }}>
      {props.children}
    </HistoryContext.Provider>
  );

}
 
export default HistoryContextProvider;