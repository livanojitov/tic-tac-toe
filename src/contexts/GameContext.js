import React, { createContext, useContext, useEffect, useReducer } from "react";
import { CategoryContext } from "./CategoryContext";
import { gameReducer } from "../store/game/gameReducer";
import { USER, RANDOM, IMAGES_PER_CATEGORY } from "../Constants";
import { changeCategory } from "../store/game/gameActions";

export const GameContext = createContext();

const initialState = {
  category: {
    id: 5,
    folder: "soccer",
    name: "SOCCER_PLAYERS",
  },

  images: {
    user: Math.floor(Math.random() * IMAGES_PER_CATEGORY),
    computer: Math.floor(Math.random() * IMAGES_PER_CATEGORY),
  },

  first: USER,

  level: RANDOM,

  disabled: true,

  over: false,
};

const GameContextProvider = ({ children }) => {
  const [game, dispatch] = useReducer(gameReducer, initialState);
  const { categories } = useContext(CategoryContext);

  useEffect(() => {
    dispatch(
      changeCategory(categories[Math.floor(Math.random() * categories.length)])
    );
  }, [categories]);

  return (
    <GameContext.Provider value={{ game, dispatch }}>
      {children}
    </GameContext.Provider>
  );
};

export default GameContextProvider;
