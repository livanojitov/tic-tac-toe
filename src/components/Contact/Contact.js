import React, { useContext } from "react";
import { DictionaryContext } from "../../contexts/DictionaryContext";
import "./Contact.scss";

const Contact = () => {
  const { dictionary } = useContext(DictionaryContext);

  const contact = {
    author: "Livan Ojito Villanueva",
    email: "livanojito@gmail.com",
    github: "https://github.com/lojito",
    linkedin: "https://ca.linkedin.com/in/lov",
  };

  return dictionary ? (
    <div className="contact" data-testid="contact">
      <p>
        <span data-testid="authorDescription">{dictionary.AUTHOR} </span>
        <span data-testid="author" className="me">
          {contact.author}
        </span>
      </p>
      <p>
        <span data-testid="emailDescription">{dictionary.EMAIL} </span>
        <span data-testid="email" className="me">
          {contact.email}
        </span>
      </p>
      <p>
        <span data-testid="githubDescription">{dictionary.GITHUB} </span>
        <a
          data-testid="github"
          target="_blank"
          href={contact.github}
          rel="noopener noreferrer"
        >
          {contact.github}
        </a>
      </p>
      <p>
        <span data-testid="linkedinDescription">{dictionary.LINKEDIN} </span>
        <a
          data-testid="linkedin"
          target="_blank"
          href={contact.linkedin}
          rel="noopener noreferrer"
        >
          {contact.linkedin}
        </a>
      </p>
    </div>
  ) : (
    <div className="loading" data-testid="loading">
      Loading the Contact page...
    </div>
  );
};

export default Contact;
