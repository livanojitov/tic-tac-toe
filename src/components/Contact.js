import React, { useState, useContext, useEffect } from 'react';
import { LanguageContext }   from '../contexts/LanguageContext';
import { HistoryContext }    from '../contexts/HistoryContext';
import { DictionaryContext } from '../contexts/DictionaryContext';

const Contact = (props) => {
  const { getLanguage } = useContext(LanguageContext);
  const language = getLanguage();
  const { changeHistory } = useContext(HistoryContext);
  const { DICTIONARY } = useContext(DictionaryContext);

  const [contact] = useState({
    author:   'Livan Ojito Villanueva',
    email :   'livanojito@gmail.com',
    github:   'https://github.com/lojito',
    linkedin: 'https://ca.linkedin.com/in/livan-ojito-villanueva'
  });

  useEffect(() => {
    return () => {
      if (props.history.location.pathname === '/history'){
        changeHistory(1);
      }else{
        changeHistory(0);
      }  
    }
  })

  return (
    <div className="contact">
      <p>
          <span>{DICTIONARY && DICTIONARY[language].AUTHOR} </span><span className="me">{contact.author}</span>
      </p>
      <p>
          <span>{DICTIONARY && DICTIONARY[language].EMAIL} </span><span className="me">{contact.email}</span>
      </p>                  
      <p>
          <span>{DICTIONARY && DICTIONARY[language].GITHUB} </span>
          <a target="_blank" href={contact.github} rel="noopener noreferrer">{contact.github}</a>
      </p>        
      <p>
          <span>{DICTIONARY && DICTIONARY[language].LINKEDIN} </span>
          <a target="_blank" href={contact.linkedin} rel="noopener noreferrer">{contact.linkedin}</a>
      </p>    
      <p>
          <span>{DICTIONARY && DICTIONARY[language].AVAILABILITY} </span>
          <span className="availability">{DICTIONARY && DICTIONARY[language].AVAILABILITY_STATUS} </span>
      </p>                  
    </div>
  )
} 

export default Contact;