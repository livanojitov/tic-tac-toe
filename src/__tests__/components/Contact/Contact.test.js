import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { DictionaryContext } from "../../../contexts/DictionaryContext";
import Contact from "../../../components/Contact/Contact";

describe("Contact", () => {
  let dictionary;
  let error;

  beforeEach(() => {
    dictionary = {
      AUTHOR: "Author:",
      EMAIL: "E-mail:",
      GITHUB: "Github:",
      LINKEDIN: "LinkedIn:",
    };
    error = "";
  });

  const renderContact = () => {
    return render(
      <DictionaryContext.Provider value={{ dictionary, error }}>
        <Contact />
      </DictionaryContext.Provider>
    );
  };

  it("renders the Contact component correctly", () => {
    const { asFragment } = renderContact();

    expect(asFragment()).toMatchSnapshot();
  });

  it("should not display the Contact info if dictionary hasn't been fetched yet", () => {
    dictionary = "";

    renderContact();

    expect(screen.queryByTestId("contact")).toBeNull();
  });

  it("should display a loading message", () => {
    dictionary = "";

    renderContact();

    expect(screen.getByTestId("loading")).toHaveTextContent(
      "Loading the Contact page..."
    );
  });

  it("should display the Contact info and apply the 'contact' class once the dictionary has been fetched", () => {
    renderContact();

    const contact = screen.getByTestId("contact");
    expect(contact).not.toBeNull();
    expect(contact).toHaveClass("contact");
  });

  it("should display the labels Author, E-mail, Github and LinkedIN in the chosen language", () => {
    renderContact();

    expect(screen.getByTestId("authorDescription")).toHaveTextContent(
      dictionary["AUTHOR"]
    );
    expect(screen.getByTestId("emailDescription")).toHaveTextContent(
      dictionary["EMAIL"]
    );
    expect(screen.getByTestId("githubDescription")).toHaveTextContent(
      dictionary["GITHUB"]
    );
    expect(screen.getByTestId("linkedinDescription")).toHaveTextContent(
      dictionary["LINKEDIN"]
    );
  });

  it("should display the contact info", () => {
    const contact = {
      author: "Livan Ojito Villanueva",
      email: "livanojito@gmail.com",
      github: "https://github.com/lojito",
      linkedin: "https://ca.linkedin.com/in/lov",
    };

    renderContact();

    expect(screen.getByTestId("author")).toHaveTextContent(contact.author);
    expect(screen.getByTestId("email")).toHaveTextContent(contact.email);
    expect(screen.getByTestId("github")).toHaveTextContent(contact.github);
    expect(screen.getByTestId("linkedin")).toHaveTextContent(contact.linkedin);
  });
});
