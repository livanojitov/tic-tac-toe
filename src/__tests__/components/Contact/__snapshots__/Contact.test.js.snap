// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact renders the Contact component correctly 1`] = `
<DocumentFragment>
  <div
    class="contact"
    data-testid="contact"
  >
    <p>
      <span
        data-testid="authorDescription"
      >
        Author: 
      </span>
      <span
        class="me"
        data-testid="author"
      >
        Livan Ojito Villanueva
      </span>
    </p>
    <p>
      <span
        data-testid="emailDescription"
      >
        E-mail: 
      </span>
      <span
        class="me"
        data-testid="email"
      >
        livanojito@gmail.com
      </span>
    </p>
    <p>
      <span
        data-testid="githubDescription"
      >
        Github: 
      </span>
      <a
        data-testid="github"
        href="https://github.com/lojito"
        rel="noopener noreferrer"
        target="_blank"
      >
        https://github.com/lojito
      </a>
    </p>
    <p>
      <span
        data-testid="linkedinDescription"
      >
        LinkedIn: 
      </span>
      <a
        data-testid="linkedin"
        href="https://ca.linkedin.com/in/lov"
        rel="noopener noreferrer"
        target="_blank"
      >
        https://ca.linkedin.com/in/lov
      </a>
    </p>
  </div>
</DocumentFragment>
`;
